src/state.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The module for managing charm state."""
+
+import logging
+import os
+import typing
+from dataclasses import dataclass
+
+import ops
+from pydantic import BaseModel, Field, ValidationError
+
+import metadata
+import server
+
+# relation name used for compatibility with machine Jenkins server charm.
+SLAVE_RELATION = "slave"
+# agent relation name
+AGENT_RELATION = "agent"
+
+logger = logging.getLogger()
+
+
+class CharmStateBaseError(Exception):
+    """Represents error with charm state."""
+
+
+class InvalidStateError(CharmStateBaseError):
+    """Exception raised when state configuration is invalid."""
+
+    def __init__(self, msg: str = ""):
+        """Initialize a new instance of the InvalidStateError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+class JenkinsConfig(BaseModel):
+    """The Jenkins config from juju config values.
+
+    Attrs:
+        server_url: The Jenkins server url.
+        agent_name_token_pairs: Jenkins agent names paired with corresponding token value.
+    """
+
+    server_url: str = Field(..., min_length=1)
+    agent_name_token_pairs: typing.List[typing.Tuple[str, str]] = Field(..., min_items=1)
+
+    @classmethod
+    def from_charm_config(cls, config: ops.ConfigData) -> typing.Optional["JenkinsConfig"]:
+        """Instantiate JenkinsConfig from charm config.
+
+        Args:
+            config: Charm configuration data.
+
+        Returns:
+            JenkinsConfig if configuration exists, None otherwise.
+        """
+        server_url = config.get("jenkins_url", None)
+        agent_name_config = config.get("jenkins_agent_name", None)
+        agent_token_config = config.get("jenkins_agent_token", None)
```

Comment:
nit comment: The `None` is not necessary, since it's `None` by default.

Reply:
That's true, thanks for the review! Fixed :D